File: nupkgs/OpenSync.Sdk/SyncMap.cs
using OpenSync.Sdk.Events;

namespace OpenSync.Sdk;

public class SyncMap
{
    private readonly OpenSyncClient _client;
    private readonly string _uniqueName;

    public event EventHandler<ItemChangedEventArgs>? OnItemAdded;
    public event EventHandler<ItemChangedEventArgs>? OnItemUpdated;
    public event EventHandler<ItemChangedEventArgs>? OnItemRemoved;

    public SyncMap(OpenSyncClient client, string uniqueName)
    {
        _client = client;
        _uniqueName = uniqueName;
        _client.OnEvent += HandleEvent;
    }

    private void HandleEvent(object? sender, SyncEventArgs args)
    {
        if (args.ObjectType == "map" && args.ObjectId == _uniqueName)
        {
            switch (args.Event)
            {
                case "item_added": OnItemAdded?.Invoke(this, new ItemChangedEventArgs(args.Data)); break;
                case "item_updated": OnItemUpdated?.Invoke(this, new ItemChangedEventArgs(args.Data)); break;
                case "item_removed": OnItemRemoved?.Invoke(this, new ItemChangedEventArgs(args.Data)); break;
            }
        }
    }

    public async Task SubscribeAsync() => await _client.SubscribeAsync("map", _uniqueName);
}

File: clients/OpenSync.Client.Js/src/SyncMap.ts
import { OpenSyncClient, SyncEvent } from './OpenSyncClient';

export class SyncMap {
  private client: OpenSyncClient;
  private name: string;

  constructor(client: OpenSyncClient, name: string) {
    this.client = client;
    this.name = name;
  }

  async subscribe(): Promise<void> {
    await this.client.subscribe('map', this.name);
  }

  onItemAdded(handler: (data: any) => void): void {
    this.client.on('item_added', (event: SyncEvent) => {
      if (event.object_id === this.name) handler(event.data);
    });
  }

  onItemUpdated(handler: (data: any) => void): void {
    this.client.on('item_updated', (event: SyncEvent) => {
      if (event.object_id === this.name) handler(event.data);
    });
  }

  onItemRemoved(handler: (data: any) => void): void {
    this.client.on('item_removed', (event: SyncEvent) => {
      if (event.object_id === this.name) handler(event.data);
    });
  }
}
